packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -118,18 +124,24 @@ export function compileBlueprint(
 		run: async (playground: UniversalPHP) => {
 			try {
 				// Start resolving resources early
-				for (const { resources } of compiled) {
+				for (const { resources, step } of compiled) {
 					for (const resource of resources) {
 						resource.setPlayground(playground);
 						if (resource.isAsync) {
-							resource.resolve();
+							resource.resolve().catch(function (error) {
```

Comment:
is there an existing bug here where we should be awaiting the response from `.resolve()`? It looks like we may continue to lines 130-144 below before all the resources have resolved.

this may be correct, as we want to queue up all the async resources to start downloading in parallel, but we don't wait for the batch of them to complete.

I guess in `resolveArguments` we end up explicitly awaiting the results so that comes through. hm. this could use a comment at a minimum.

it seems that the `onStepError` callback might execute in the middle of processing or waiting for other resources. that also seems like it's probably fine, but possibly surprising.

in other words, this doesn't particularly give us an opportunity to choose to abort initializing the playground if a step fails even though it gives us the chance to know _while_ initializing if one or more things fail.

---

as someone using the Playground I wonder what I care about:
 - I don't care if one plugin fails to load as long as I can know it failed. keep going.
 - if my one particular thing fails then there's no point continuing.

in the end, I guess there's no harm in doing a partial boot of the playground as long as we're not connected to an OPFS or Live Directory share. partial boot in those places could corrupt files, though I hope people aren't relying on this _not_ corrupting files. it's supposed to be ephemeral.

---

did you consider passing a callback when we finish running the steps?

the current callbacks speak to individual steps, but not to the overall progress.

```ts
const progressObject = new Map<StepDefinition, Promise<void>>();
const progressPromises = new Map<StepDefinition, {resolve: () => void, reject: (error: string) => void}>();
for (const { step } of compiled) {
	progressObject.set(step, new Promise((resolve, reject) => {
		progressPromises.set(step, {resolve, reject});
	} ) );
}
onRunningSteps( progressObject );
```

then when resolving we can do this…
```ts
	for (const { run, step } of compiled) {
		try {
			const result = await run(playground);
			progressPromises.get(step).resolve();
		} catch (error) {
			progressPromises.get(step).reject();
		}
	}
} finally {
	try {
		if (onStepsCompleted) {
			await onStepsCompleted( steps );
		}
		await (playground as any).goTo(
			blueprint.landingPage || '/'
		);
	} catch (e) {
		/*
		 * NodePHP exposes no goTo method.
		 * We can't use `goto` in playground here,
		 * because it may be a Comlink proxy object
		 * with no such method.
		 */
	}
	progress.finish();
}
```

Implementors could still replicate the behavior that `onStepCompleted` and `onStepError` provide, but they could additional prevent the boot-up of the playground.

```ts
// Prevent booting if anything fails.
onStepsCompleted = async ( steps ) => {
	const results = await Promise.allSettled( steps.values() );
	if ( results.any( p => p.status === 'rejected' ) ) {
		throw new Error( "Don't boot unless everything loads." );
	}
}
```

or report in the DOM the failures

```ts
// Announce the failure, but keep on going.
onRunningSteps = steps => {
	steps.forEach( ( promise, step ) => {
		promise.catch( () => {
			const alert = document.createElement( 'div' );
			alert.classList.add( 'alert' );
			alert.innerHTML = `Failed to load '${ step.step }'`.
			document.querySelector( 'some selector' ).appendChild( alert );
		} );
	} );
}
```

obviously each of these could be inspected for better reporting or logic. in fact, maybe `onRunningSteps` isn't even as useful as `onStepsCompleted`.

Reply:
> Is there an existing bug here where we should be awaiting the response from .resolve()? It looks like we may continue to lines 130-144 below before all the resources have resolved.

It's an unusual way to use Promises, to instantiate all of them without await, then, or catch. I had to think about this part when I read it first, but I understand it's intentional in this case, to start downloading resources from all steps in parallel, before executing any step. Internally, `resource.resolve()` uses a "semaphore" to manage a queue with a maximum number of concurrency.

> This may be correct, as we want to queue up all the async resources to start downloading in parallel, but we don't wait for the batch of them to complete.

A possible buggy part of this logic was that there was no `catch()` when a resource promise threw an error. It would show up in the browser dev console as an uncaught promise. With the change in this pull request, such errors are caught and passed to the `onStepError` callback, if any, or rethrown as an error of the async `CompiledBlueprint.run()` function for the caller to handle.

---

> I guess in `resolveArguments` we end up explicitly awaiting the results so that comes through. hm. this could use a comment at a minimum.

OK, I'll add:

```ts
/**
 * When each step's `CompiledStep.run()` is called, it uses
 * `resolveArguments` to await all dependent resources for the
 * step.
 */
```

Do you think the wording is clear enough?

---

> It seems that the onStepError callback might execute in the middle of processing or waiting for other resources. that also seems like it's probably fine, but possibly surprising.

That's true, the `onStepError` callback is called when `resource.resolve()` fails, or when `CompiledStep.run()` fails.

But in practice I think it's OK, because the callback receives the error as well as the step wherein it occurred, and a resource fetch can be considered part of the step: if it fails, the step fails, and the callback can decide if that's a show-stopper or to let the site load.

---

> - I don't care if one plugin fails to load as long as I can know it failed. keep going.
> - if my one particular thing fails then there's no point continuing.

Hmm, I lean toward the latter approach, I'd prefer to stop the whole thing if a plugin that I specified fails to load. Having a site in partially incomplete state seems like it would lead to unexpected errors down the road, for example in other steps that rely on a non-existent plugin.

That's how it worked originally, though without any error message to the user, I think, or it flashed an error in the progress screen, then continued loading the site without the plugin.

With the current state of this PR, there's no option in the Query API to specify that a plugin is *required* instead of optional, and to prevent loading the site if a plugin install fails.

In other words, I think we may need the opposite of the feature that the user was suggesting in #600, [An invalid plugin slug should not cause following plugins to not be installed](https://github.com/WordPress/wordpress-playground/issues/600). It's probably desirable in some situations, for an invalid plugin slug or plugin install error to stop the whole thing, rather than continue with steps that require it.

> in the end, I guess there's no harm in doing a partial boot of the playground as long as we're not connected to an OPFS or Live Directory share. partial boot in those places could corrupt files, though I hope people aren't relying on this not corrupting files. it's supposed to be ephemeral.

If a plugin is considered optional ("Let the site load even if it fails to install"), then I suppose it should be fine even with a persistent site. Who knows, the next time the plugin may successfully load! (Which is an approach I'm not that comfortable with, but apparently in some situations that is the desired behavior.)

---

> did you consider passing a callback when we finish running the steps? the current callbacks speak to individual steps, but not to the overall progress.

I hadn't considered it, but at first glance it seems like `onStepsCompleted` could simply be when the function returns? Or when it throw an error, instead of an `onStepsError` callback.

> Implementors could still replicate the behavior that `onStepCompleted` and `onStepError` provide, but they could additional prevent the boot-up of the playground.

Oh I see, you mean a different solution: instead of these step-specific callbacks, to have a single callback for all steps.

Hmm, but that approach would change the default behavior of throwing an error when any step fails (including steps to install plugins)?

I think the existing behavior ("Stop everything if a step or plugin install fails") is more intuitive, and that a fail-able step or an optional plugin is the exception, better suited as an opt-in behavior?

> obviously each of these could be inspected for better reporting or logic. in fact, maybe onRunningSteps isn't even as useful as onStepsCompleted.

How about, instead of an `onStepsCompleted` callback, the function `CompiledBlueprint.run()` could `return await Promise.allSettled( steps.values() )`? That could include return values of each step on success, or any error per step.

Well, I imagine it's do-able either way, to let the caller handle each error per step, or all step errors per blueprint. For example, the `onStepError` callback could display an alert and let the remaining steps continue or not.

Reply:
> we may need the opposite of the feature that the user was suggesting in https://github.com/WordPress/wordpress-playground/issues/600, https://github.com/WordPress/wordpress-playground/issues/600. It's probably desirable in some situations, for an invalid plugin slug or plugin install error to stop the whole thing, rather than continue with steps that require it.

To elaborate, for the Query API, perhaps each specified plugin should be required by default (stop everything if a plugin install fails), instead of optional by default as in this PR (let the site load even if a plugin install fails).

Outside of the Query API, I think it makes sense for `CompiledBlueprint.run()` to throw if any step or plugin install fails, as default behavior, and optionally let it be fail-able.

Reply:
> It's an unusual way to use Promises

it's not that unusual, especially in situations like this, maybe just less common than when working in async flows where we're waiting on individual promises in a chain of execution.

> onStepsCompleted could simply be when the function returns?

I like this idea, as long as we can signal back whether execution should proceed.

> Hmm, I lean toward the latter approach, I'd prefer to stop the whole thing if a plugin that I specified fails to load. Having a site in partially incomplete state seems like it would lead to unexpected errors down the road, for example in other steps that rely on a non-existent plugin.

here I wasn't tossing around two options we should choose from, but trying to list legitimate cases people will want. I don't think either one is a right choice but rather I think there will be plenty of situations where someone wants to make that decision themself, and might even make both decisions in a single playground. for example, _this_ plugin is a show-stopped but the theme is something I don't care about, or the caching plugin.

it's more about giving control to people building sites or tools with the Playground. the idea is that if we can report the progress or completion status then they can indicate whether to boot up the Playground or terminate with an error.

> that approach would change the default behavior of throwing an error when any step fails (including steps to install plugins)?

yeah, and I wouldn't think it's as important if we signal this at every step as it fails or if we wait until the end and list all the completion statuses. wouldn't it only change the behavior from what's proposed already in this PR, though, not how it works today? (since today it doesn't report any errors)

> If a plugin is considered optional ("Let the site load even if it fails to install"), then I suppose it should be fine even with a persistent site. Who knows, the next time the plugin may successfully load! (Which is an approach I'm not that comfortable with, but apparently in some situations that is the desired behavior.)

right, and that's why I think a few people have reported back that they want the Playground to continue booting when something doesn't load. the important bits are there, the extras aren't.

if the right steps are required then it's fine on a persisted site, but if, for example, some plugin transforms data on save or maintains table relationships, and it doesn't load for a persistent site, then we could get into situations where it corrupts the database or file uploads just because the site continues to run without part of its functionality. in this case that would be caused by someone not instructing the thing to fail on error, but it would be something we would want to make sure we communicate could happen.

I only wanted to mention that there's this difference in risk for temporary sites vs. those persisting their data, particularly those using the Live Directory in Chrome or using `wp-now` or otherwise using the Playground while connected to real data.

Reply:
I can try to rewrite the logic and implement `onStepsCompleted` instead of `onStepError` and `onStepCompleted` (which was already there before this PR) - but I'm not sure I understand the proposed change.

```ts
	for (const { run, step } of compiled) {
		try {
			const result = await run(playground);
			progressPromises.get(step).resolve();
		} catch (error) {
			progressPromises.get(step).reject();
		}
	}
} finally {
	try {
		if (onStepsCompleted) {
			await onStepsCompleted( steps );
		}
```

Within the catch clause for each step, if it only sets the rejected status for each promise, it seems there isn't an opportunity for the caller of `CompiledBlueprint.run()` (such as the website or on server side) to abort the process, unlike with the `onStepError` callback? And a failed step would always let remaining steps continue?

..Ah I see, that's the purpose of the `onRunningSteps` callback. It loops through the step promises beforehand, so it can attach `catch()` before starting the execution of steps. And if there is no `onRunningSteps` callback passed, then a failed step promise will let other steps continue. Such failed steps are passed to `onStepsCompleted()`, which can choose to continue loading the site (and call progress.finish()), or stop and show an error.

OK, I think I understand the intent. I can rewrite the logic in this way.

One point I'm not so convinced about, that the default behavior of a failed step should be to continue running the rest of the steps. Unlike on the Playground website, when using the Blueprints API from the Playground Client module or from Node.js like with `wp-now`, I think it's more intuitive for a step error to be a show-stopper by default?

Reply:
> I only wanted to mention that there's this difference in risk for temporary sites vs. those persisting their data, particularly those using the Live Directory in Chrome or using wp-now or otherwise using the Playground while connected to real data.

Right, I see what you mean about the risk of corrupting data. This reenforces my hesitation to let a site continue loading in partial/incomplete state, i.e., when a required plugin is not installed. I think it should be opt-in.

Just as an idea, if we decide to let plugin installs fail by default for the Query API - maybe there could be a syntax to append or prepend `!` (exclamation mark) to the plugin slug to mark it as required.

Reply:
OK, here's how it looks with the logic rewritten, changed from using `onStepCompleted` and `onStepError` to using `onRunningSteps` and `onStepsCompleted`.

```ts
const progressMap = new Map<StepDefinition, Promise<void>>();
const progressPromises = new Map<StepDefinition, {resolve: () => void, reject: (error: string) => void}>();
for (const { step } of compiled) {
	progressMap.set(step, new Promise((resolve, reject) => {
		progressPromises.set(step, {resolve, reject});
	} ) );
}
onRunningSteps( progressMap );

try {
	// Start resolving resources early
	for (const { resources, step } of compiled) {
		for (const resource of resources) {
			resource.setPlayground(playground);
			if (resource.isAsync) {
				resource.resolve().catch(function (error) {
					// onStepError(error, step);
					progressPromises.get(step).reject(error);
				});
			}
		}
	}

	/**
	 * When each compiled step's `run` method is called, it uses
	 * `resolveArguments` to await dependent resources for the step.
	 */
	for (const { run, step } of compiled) {
		try {
			const result = await run(playground);
			// onStepCompleted(result, step);
			progressPromises.get(step).resolve(result);
		} catch (error) {
			// onStepError(error, step);
			progressPromises.get(step).reject(error);
		}
	}
} finally {
	try {
		if (onStepsCompleted) {
			await onStepsCompleted( progressMap );
		}
		await (playground as any).goTo(
			blueprint.landingPage || '/'
		);
	} catch (e) {
		/*
			* NodePHP exposes no goTo method.
			* We can't use `goto` in playground here,
			* because it may be a Comlink proxy object
			* with no such method.
			*/
	}
	progress.finish();
}
```

---

Perhaps `onRunningSteps` could be called `beforeRunningSteps` to clarify when it is called.

The new logic will accomplish basically the same thing as per-step callbacks, namely to provide the caller an ability to interrupt or continue running the ramaining steps when a step fails.

A possible disadvantage is that it's a bit more complicated for the caller. They must know how the `onRunningSteps` callback is meant to work, by looping through promises to attach `catch` (and maybe `then`?) callbacks. I think per-step callbacks are easier to understand and use, by being passed the step result or error directly.

One advantage of the per-blueprint callbacks that I can see, is that in addition to the ability to interrupt/continue steps after a step fails, it gives the ability after all steps have finished running, to decide whether to continue loading the site with `playground.goTo( landingPage )` based on the step results. (But the `onStepError` callback can do that too, just not at the end but when an error occurs.)

Reply:
> right, and that's why I think a few people have reported back that they want the Playground to continue booting when something doesn't load. the important bits are there, the extras aren't.

So, with the current state of this PR being "Continue plugin installs on error", we're letting the remaining steps continue and the site to load, even when any plugin fails to install. In effect, for the website Query API, we would be considering all plugins to be optional ("extras") and not strictly required ("important"). I wonder if that's what we want.

For the Query API, I think it does make sense in terms of user experience, to let the website load even if a plugin install fails. That's far better than getting an error screen and being unable to play with the site at all, even if it's partial/incomplete.

For all other uses of the Blueprint module, via Playground Client or on the server side, I think it's more intuitive and common practice to throw any error as soon as it occurs; and only optionally let the caller decide if a step is optional and fail-able.

Reply:
> One point I'm not so convinced about, that the default behavior of a failed step should be to continue running the rest of the steps…it's more intuitive for a step error to be a show-stopper by default?

that's fine because I'm not trying to convince of that 😄 I just want to make it possible for someone to determine that, and right now without your awesome contribution here it's just not possible. so I'm not arguing to keep going; I'm only arguing to build this system which lets _other people_ choose, and specifically avoiding _us_ making that call for them.

> maybe there could be a syntax to append or prepend ! (exclamation mark) to the plugin slug to mark it as required.
> (..Meh, on second thought, that's not that great, it will be another thing people would have to learn. The Query API is already a bit complex to use

I concur with you here and I think that if someone wants something more complicated they can already do that. in follow-up work I think we could look at the plugin and theme blueprint steps and add an `optional` flag to indicate that it's fine to continue if they fail. this would be the easy canonical way for people to set that up, and it would be a short jump from calling `?plugin=my-plugin` to adding `#{"steps":[{"step":"installPlugin", "pluginName": "my-plugin"}]}` (yes it's verbose and yes I think there is another step involved here in getting the plugin zip, but these things can also be handled through a better UI and documentation - a blueprint fragment builder of sorts, exposing `compileBlueprints` yada yada).

---

concerning the callback maybe we have some iteration to do on the design.

> to abort the process, unlike with the onStepError callback? And a failed step would always let remaining steps continue?

okay I think I am conflating two things: steps and resources. resources can load concurrently but steps are sequenced.

```js
// Start resolving resources early
for (const { resources, step } of compiled) {
	for (const resource of resources) {
		resource.setPlayground(playground);
		if (resource.isAsync) {
			resource.resolve().catch(function (error) {
				// onStepError(error, step);
				progressPromises.get(step).reject(error);
			});
		}
	}
}
```

in this chunk we start fetching all resources concurrently. we have no idea in what order they are going to resolve, so the `onStepError` callback at this point doesn't provide much more than a single callback after all resources are resolved. yes, we may call it as soon as the first dependent resource fails for a given step, but that could also be the last of many resources to fetch. we can imagine such a scenario when we're installing a plugin and a theme. even if the plugin is the first step, the server could be slow sending the ZIP and the theme resource might appear first. in this case, by the time the plugin "fails" the theme resources are already here.

<img src="https://japari-library.com/w/images/0/0c/WhackAMoleUniqueBehavior.gif" align=right>

this is why I first thought of the promise map, because I imagined something like a whack-a-mole board where steps settle as their dependent resources fully succeed or fail. this is also why I was first confused by `resolveArguments` because we really don't know after this nested for loop if a step is ready, we have to wait for `resolveArguments` (or check the status of each promise in the resources for that step). that could involve more than fetches too, as dependent directory creation or file copying could also fail.

though I didn't show it in my code snippet, I was thinking for the `onStepsCompleted` callback, something like this would work

```js
if ( onStepsCompleted && await onStepsCompleted( progressMap ) ) {
	await playground.goTo( blueprint.landingPage || '/' );
} else {
	showErrorScreen();
}
```

we don't have to expose the promise map at this point either: we could reduce it to a boolean status of step -> status

first, a tiny helper that I'll use a couple of times here

```ts
/**
 * Convert a Promise into a boolean indicating if it resolved or rejeted.
 */
const didResolve = <P extends Promise<any>>(promise: P): bool =>
	p.then( () => true, () => false );
```

```js
const completedStatus = new Map();
for ( const { step } of compiled ) {
	completedStatus.set(
		step,
		await didResolve( progressMap.get( step ) )
	);
}
onStepsCompleted( completedStatus );
```

my earlier ideas about `onRunningSteps` was that it would give the caller more control to totally abort everything once something fails, as in, to stop downloading a large dependent ZIP file.

```js
onRunningSteps = ( steps, { abortController } ) => {
	// Without my plugin, nothing should boot up.
	steps
		.find( ( { step, pluginName } ) => step === 'activatePlugin' && pluginName === 'my-important-plugin' )
		.catch( () => abortController.abort() );
}
```

maybe that's more of an enhancement than a necessary part of this. it all still gives a place to error out immediately once a step fails, but only if someone wants that. so maybe the default should finally look something like this:

```js
// Either someone made a choice to boot or all steps succeeded.
const shouldBoot = onStepsCompleted
	? await onStepsCompleted( completedStatus )
	: await didResolve( Promise.all( [ ...progressMap.values ] ) );

if ( shouldBoot ) {
	await playground.goTo( blueprint.landingPage || '/' );
} else {
	showErrorScreen();
}
```

with this, all someone needs to do is write a function that examines the steps and a `true` or `false` value if they succeeded, and return `true` or `false` to indicate if the Playground should proceed.

if we add `optional` to plugin and theme Blueprint steps, all we have to do is implement this ourself with a custom `onStepsCompleted` handler.

thanks for going through all this discussion and back and forth. I think it's good that we struggle with it now rather than after we publish and people start building on it.

Reply:
- Commit 1281355: [Implement `onRunningSteps` for success/error handling per step; and `onStepsCompleted` for success/error handling per blueprint](https://github.com/WordPress/wordpress-playground/commit/1281355bbb992e74245a55e581ed1249d71e20c2)

Reply:
noice. I'm going to mark this thread resolved because it's getting in the way with its length. feel free to unresolve it, or open a new thread to continue chatting about it.